File: 0739-daily-temperatures/0739-daily-temperatures.cpp

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        int n = T.size();
        vector<int> result(n, 0); // Initialize result vector with zeroes
        stack<int> st; // Stack to store indices of temperatures

        for (int i = 0; i < n; ++i) {
            // While stack is not empty and current temperature is greater than temperature at index of top of stack
            while (!st.empty() && T[i] > T[st.top()]) {
                int idx = st.top();
                st.pop(); // Get the index from the top of the stack
                result[idx] = i - idx; // Calculate the number of days until a warmer temperature
            }
            // Push the current index onto the stack
            st.push(i);
        }

        return result;
    }
};

File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, T: List[int]) -> List[int]:
            
        n = len(T)
        result = [0] * n  # Initialize result list with zeroes
        stack = []  # Stack to store indices of temperatures

        for i in range(n):
            # While stack is not empty and current temperature is greater than temperature at index of top of stack
            while stack and T[i] > T[stack[-1]]:
                idx = stack.pop()  # Get the index from the top of the stack
                result[idx] = i - idx  # Calculate the number of days until a warmer temperature
            # Push the current index onto the stack
            stack.append(i)

        return result

